src/cdn.rs

```diff
@@ -117,75 +258,202 @@ impl CdnBackend {
                     .build(),
             )
             .send()
-            .await?;
+            .await?
+            .invalidation()
+            .ok_or_else(|| {
+                anyhow!("missing invalidation information in create-invalidation result")
+            })?
+            .id()
+            .ok_or_else(|| anyhow!("missing invalidation ID in create-invalidation result"))?
+            .to_owned())
+    }
+}
 
-        Ok(())
+#[instrument(skip(conn))]
+pub(crate) fn handle_queued_invalidation_requests(
+    cdn: &CdnBackend,
+    conn: &mut impl postgres::GenericClient,
+    distribution_id: &str,
+) -> Result<()> {
+    info!("handling queued CDN invalidations");
+
+    let active_invalidations = cdn
+        .active_invalidations(distribution_id)
+        .context("error querying active invalidations")?;
+
+    // for now we assume all invalidation paths are wildcard invalidations,
+    // so we apply the wildcard limit.
+    let active_path_invalidations: usize = active_invalidations
+        .iter()
+        .map(|i| i.path_patterns.len())
+        .sum();
+
+    debug!(
+        active_invalidations = active_invalidations.len(),
+        active_path_invalidations, "found active invalidations",
+    );
+
+    // remove the invalidation from the queue when they are completed.
+    // We're only looking at InProgress invalidations,
+    // we don't differentiate between `Completed` ones, and invalidations
+    // missing in the CloudFront `ListInvalidations` response.
+    conn.execute(
+        "DELETE FROM cdn_invalidation_queue
+         WHERE 
+             cdn_distribution_id = $1 AND 
+             created_in_cdn IS NOT NULL AND 
+             NOT (cdn_reference = ANY($2))",
+        &[
+            &distribution_id,
+            &active_invalidations
+                .iter()
+                .map(|i| i.invalidation_id.clone())
+                .collect::<Vec<_>>(),
+        ],
+    )?;
+
+    let possible_path_invalidations: i32 =
+        MAX_CLOUDFRONT_WILDCARD_INVALIDATIONS - active_path_invalidations as i32;
+
+    if possible_path_invalidations <= 0 {
+        info!(
+            active_path_invalidations,
+            "too many active cloudfront wildcard invalidations \
+            will not create a new one."
+        );
+        return Ok(());
+    }
+
+    // create new an invalidation for the queued path patterns
+    let mut transaction = conn.transaction()?;
+    let mut path_patterns: Vec<String> = Vec::new();
+    let mut queued_entry_ids: Vec<i64> = Vec::new();
+
+    for row in transaction.query(
+        "SELECT id, path_pattern
+         FROM cdn_invalidation_queue 
+         WHERE cdn_distribution_id = $1 AND created_in_cdn IS NULL
+         ORDER BY queued, id
```

Comment:
Why order by `id` as well as `queued`? It seems like the queued timestamp should be near enough unique that having a second clause in the ORDER won't matter.

I think you need the index on `created_in_cdn` to be instead an index on (`created_in_cdn`, `queued`) in order to make this query cheap.

Reply:
I added ordering by id because I'm used to it to guarantee that the order is deterministic in any case, even when we manually change any data in the queue. 

About the index you're right in general, not sure if it makes that big of a difference with the estimated queue lengths.